xlate/openroadm: initial support

KazuyaAnazawa opened the pull request:
This PR adds OpenROADM (v.10) translator.
It supports to translate betweeen the following OpenROADM YANG models and Goldstone YANG models:
- `org-openroadm-device` model
- `org-openroadm-pm` model

This implementation is contributed from Fujitsu and NTT.

KazuyaAnazawa commented:
The CI fails at lint stage, so I will look into and fix it.

KazuyaAnazawa commented:
@ishidawataru 
As for the errors at lint stage, I found that several source codes under `src` should be reformatted.
So I pushed a commit https://github.com/oopt-goldstone/goldstone-mgmt/pull/74/commits/29eb2e97b7a55408c55d2243464c4ad4a97ea540, and now CI passed.
But if it's better to create another branch and PR for this, please let me know. I will do it.

For more details, I would like to share what I did.
1. Reproduced the error at lint stage in my local environment.
```
# docker run -t -v `pwd`:`pwd` -w `pwd` ghcr.io/oopt-goldstone/mgmt/tester:latest make lint
which black && exit `black -q --diff --exclude "src/north/snmp/src|src/north/gnmi/goldstone/north/gnmi/proto" src | wc -l`
/usr/local/bin/black
make: *** [Makefile:94: lint] Error 73
```

2. Reformatted the codes as follows.
```
# docker run -t -v `pwd`:`pwd` -w `pwd` ghcr.io/oopt-goldstone/mgmt/tester:latest black --exclude "src/north/snmp/src|src/north/gnmi/goldstone/north/gnmi/proto" src
reformatted /home/ntt/oss-goldstone-mgmt/src/north/cli/tests/test_system.py
reformatted /home/ntt/oss-goldstone-mgmt/src/lib/tests/test_abort.py
reformatted /home/ntt/oss-goldstone-mgmt/src/north/cli/tests/test_management_interfaces.py
reformatted /home/ntt/oss-goldstone-mgmt/src/north/notif/gsnotif/main.py
reformatted /home/ntt/oss-goldstone-mgmt/src/lib/goldstone/lib/connector/base.py
reformatted /home/ntt/oss-goldstone-mgmt/src/lib/tests/test_sysrepo.py
reformatted /home/ntt/oss-goldstone-mgmt/src/north/cli/goldstone/north/cli/system.py
reformatted /home/ntt/oss-goldstone-mgmt/src/north/cli/tests/test_transponder.py
reformatted /home/ntt/oss-goldstone-mgmt/src/south/sonic/goldstone/south/sonic/__init__.py
reformatted /home/ntt/oss-goldstone-mgmt/src/lib/goldstone/lib/server_connector/sysrepo.py
reformatted /home/ntt/oss-goldstone-mgmt/src/lib/goldstone/lib/core.py
reformatted /home/ntt/oss-goldstone-mgmt/src/north/cli/goldstone/north/cli/main.py
reformatted /home/ntt/oss-goldstone-mgmt/src/xlate/openconfig/goldstone/xlate/openconfig/__init__.py
reformatted /home/ntt/oss-goldstone-mgmt/src/south/system/gssystem/k8s.py
reformatted /home/ntt/oss-goldstone-mgmt/src/north/cli/goldstone/north/cli/platform.py
reformatted /home/ntt/oss-goldstone-mgmt/src/north/cli/goldstone/north/cli/portchannel.py
reformatted /home/ntt/oss-goldstone-mgmt/src/north/cli/tests/test_interfaces.py
reformatted /home/ntt/oss-goldstone-mgmt/src/system/telemetry/goldstone/system/telemetry/store.py
reformatted /home/ntt/oss-goldstone-mgmt/src/north/cli/goldstone/north/cli/ufd.py
reformatted /home/ntt/oss-goldstone-mgmt/src/north/cli/goldstone/north/cli/aaa.py
reformatted /home/ntt/oss-goldstone-mgmt/src/north/cli/goldstone/north/cli/vlan.py
reformatted /home/ntt/oss-goldstone-mgmt/src/south/gearbox/goldstone/south/gearbox/gearbox.py
reformatted /home/ntt/oss-goldstone-mgmt/src/north/cli/goldstone/north/cli/base.py
reformatted /home/ntt/oss-goldstone-mgmt/src/south/sonic/goldstone/south/sonic/k8s_api.py
reformatted /home/ntt/oss-goldstone-mgmt/src/south/tai/tests/test.py
reformatted /home/ntt/oss-goldstone-mgmt/src/south/sonic/goldstone/south/sonic/interfaces.py
reformatted /home/ntt/oss-goldstone-mgmt/src/south/gearbox/tests/tests.py
reformatted /home/ntt/oss-goldstone-mgmt/src/south/tai/goldstone/south/tai/transponder.py
reformatted /home/ntt/oss-goldstone-mgmt/src/north/cli/goldstone/north/cli/interface.py

All done! ? ? ?
29 files reformatted, 121 files left unchanged.
```

ishidawataru commented on src/xlate/openroadm/goldstone/xlate/openroadm/device.py:
```diff
@@ -0,0 +1,1484 @@
+GS_TRANSPONDER_HOSTIF_FEC_TYPE = GS_TRANSPONDER_HOSTIF + "/config/fec-type"
+
+
+# OR ports are 1-16
+# GS has piu1/hi0-3 piu2/hi0-3...
+or_port_lookup = {
+    "port1": ["piu1", 0],
+    "port2": ["piu1", 1],
+    "port3": ["piu1", 2],
+    "port4": ["piu1", 3],
+    "port5": ["piu2", 0],
+    "port6": ["piu2", 1],
+    "port7": ["piu2", 2],
+    "port8": ["piu2", 3],
+    "port9": ["piu3", 0],
+    "port10": ["piu3", 1],
+    "port11": ["piu3", 2],
+    "port12": ["piu3", 3],
+    "port13": ["piu4", 0],
+    "port14": ["piu4", 1],
+    "port15": ["piu4", 2],
+    "port16": ["piu4", 3],
+}
```
This only works with GFT. Please consider a more generic way of handling this.

ishidawataru commented on src/xlate/openroadm/goldstone/xlate/openroadm/main.py:
```diff
@@ -0,0 +1,76 @@
+    else:
+        logging.basicConfig(level=logging.INFO, format=fmt)
+
+    operational_modes = load_operational_modes(args.operational_modes_file)
```
It looks like this operational mode configuration file is not used.
Please don't hardcode the mapping of mode attributes and use this configuration file to do the mapping.
Feel free to extend the format of this file as needed.

ishidawataru reviewed (changes requested):
I left two comments. Please take a look.

ishidawataru commented:
@KazuyaAnazawa Thanks. Please open a new PR to fix the CI.

KazuyaAnazawa commented:
@ishidawataru I'm so sorry for the delay.

I have fixed several parts according to the comments. Could you check this?
- I reused `platform.json` for port mapping in https://github.com/oopt-goldstone/goldstone-mgmt/pull/74/commits/05f0d61880d70e67ae03ed369b7abc225ce2e21f.
- `operational-modes.json` was used [here](https://github.com/oopt-goldstone/goldstone-mgmt/blob/xlate-or/initial-support/src/xlate/openroadm/goldstone/xlate/openroadm/device.py#L248), but only one operational mode was defined. I added more operational modes in https://github.com/oopt-goldstone/goldstone-mgmt/pull/74/commits/a99a358887d3418b6f275743b0aa90121281ddec. I have checked the modes (profiles) are correctly configured on real GFT.
- Another minor mistake is fixed in https://github.com/oopt-goldstone/goldstone-mgmt/pull/74/commits/639606293e9c9955004e8b8d97a1892eed28ecc8.

ishidawataru commented on src/xlate/openroadm/tests/platform.json:
```diff
@@ -0,0 +1,70 @@
+  {"synce-reference-clock": {"name": "1", "index": 1, "gearbox": {"name": "2"}}, "input-reference": {"name": "5", "index": 5, "dpll": {"name": "1"}}},
+  {"synce-reference-clock": {"name": "2", "index": 2, "gearbox": {"name": "2"}}, "input-reference": {"name": "6", "index": 6, "dpll": {"name": "1"}}},
+  {"synce-reference-clock": {"name": "3", "index": 3, "gearbox": {"name": "2"}}, "input-reference": {"name": "7", "index": 7, "dpll": {"name": "1"}}},
+  {"openroadm-port-map": {"openroadm-port-name": "port1", "goldstone": {"module": "piu1", "hostif": 0}}},
+  {"openroadm-port-map": {"openroadm-port-name": "port2", "goldstone": {"module": "piu1", "hostif": 1}}},
+  {"openroadm-port-map": {"openroadm-port-name": "port3", "goldstone": {"module": "piu1", "hostif": 2}}},
+  {"openroadm-port-map": {"openroadm-port-name": "port4", "goldstone": {"module": "piu1", "hostif": 3}}},
+  {"openroadm-port-map": {"openroadm-port-name": "port5", "goldstone": {"module": "piu2", "hostif": 0}}},
+  {"openroadm-port-map": {"openroadm-port-name": "port6", "goldstone": {"module": "piu2", "hostif": 1}}},
+  {"openroadm-port-map": {"openroadm-port-name": "port7", "goldstone": {"module": "piu2", "hostif": 2}}},
+  {"openroadm-port-map": {"openroadm-port-name": "port8", "goldstone": {"module": "piu2", "hostif": 3}}},
+  {"openroadm-port-map": {"openroadm-port-name": "port9", "goldstone": {"module": "piu3", "hostif": 0}}},
+  {"openroadm-port-map": {"openroadm-port-name": "port10", "goldstone": {"module": "piu3", "hostif": 1}}},
+  {"openroadm-port-map": {"openroadm-port-name": "port11", "goldstone": {"module": "piu3", "hostif": 2}}},
+  {"openroadm-port-map": {"openroadm-port-name": "port12", "goldstone": {"module": "piu3", "hostif": 3}}},
+  {"openroadm-port-map": {"openroadm-port-name": "port13", "goldstone": {"module": "piu4", "hostif": 0}}},
+  {"openroadm-port-map": {"openroadm-port-name": "port14", "goldstone": {"module": "piu4", "hostif": 1}}},
+  {"openroadm-port-map": {"openroadm-port-name": "port15", "goldstone": {"module": "piu4", "hostif": 2}}},
+  {"openroadm-port-map": {"openroadm-port-name": "port16", "goldstone": {"module": "piu4", "hostif": 3}}}
```
How about adding a field to the existing elements instead of adding new openroadm specific elements?

Something like the following.

```json
{
  "interface": {
    "name": "Interface1/1/1",
    "default": true,
    "pin-mode": "pam4"
  },
  "tai": {
    "module": {
      "name": "piu1",
      "location": "1"
    },
    "hostif": {
      "name": "0",
      "index": 0
    }
  },
  "openroadm": {
    "name": "port1"
  }
}
```

Please note that the number of interfaces may change depending on the configuration (e.g., PAM4 or NRZ).

ishidawataru reviewed (changes requested):
Thanks for the update.
I left one review comment. Please take a look.

KazuyaAnazawa commented:
@ishidawataru So sorry for delayed update.
I have fixed the source codes according to comments https://github.com/oopt-goldstone/goldstone-mgmt/pull/74#discussion_r1195824902.
Sorry to bother, but could you review fixed codes?

ishidawataru commented:
@KazuyaAnazawa Thanks. Looks good to me. Please clean up the commits.

KazuyaAnazawa commented:
@ishidawataru Thank you for your review. 
I have cleaned up the commits.
